File: cpp/0876.0_Middle_of_the_Linked_List.cpp
/*
执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：6.8 MB, 在所有 C++ 提交中击败了86.91% 的用户
通过测试用例：36 / 36
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* slow = head, * fast = head;
        while (fast != NULL && (fast->next != NULL)) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
};

File: leetcode-cn/0876.0_Middle_of_the_Linked_List.py
'''
执行用时：40 ms, 在所有 Python3 提交中击败了40.67% 的用户
内存消耗：14.8 MB, 在所有 Python3 提交中击败了89.81% 的用户
通过测试用例：36 / 36
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        slow = fast = head 
        while fast and fast.next:
            slow = slow.next 
            fast = fast.next.next 
        return slow 
        
        